src/components/PageSwitch.tsx

```diff
@@ -36,13 +40,36 @@ const PageSwitch = ({
   const currentPage = parseInt(useQueryParams('p') || '1')
   const navigate = useNavigate()
   const location = useLocation()
+  const [isSettingPageManually, setIsSettingPageManually] = useState(false)
+  const [manuallySetPage, setManuallySetPage] = useState<number | ''>(currentPage)
+
+  const manualPageInputRef = useRef<HTMLInputElement>(null)
 
   const locationSearch = useMemo(() => new URLSearchParams(location.search), [location.search])
 
+  useOnClickOutside({ ref: manualPageInputRef, handler: () => setIsSettingPageManually(false) })
+
   const handlePageSwitch = (direction: 'previous' | 'next') => {
     setPageNumber(direction === 'previous' ? currentPage - 1 : currentPage + 1)
   }
 
+  const handlePageInputChange = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.value === '') setManuallySetPage('')
+
+    const pageNumber = parseInt(e.target.value)
```

Comment:
Hm, what if the result is `NaN` or some really huge number input that cannot be parsed?